### PR TITLE
gptp: ignore syncFollowUp with negative correction field

### PR DESCRIPTION
**Pull request by PawelModrzejewski**
Simple robustness against switch/bridge issue of putting negative values in CorrectionField.

### CODE REVIEW

**Comment by PawelModrzejewski** [comment]
Any comments here?

**Comment by andrew-elder** [comment]
Sorry for the delay. This looks ok to me. Anyone else have comments.

**Comment by andrew-elder** [comment]
@christopher-s-hall - any comments?

**Comment by christopher-s-hall** [comment]
What problem does this solve? Is a negative correction field an error? Apart from that, the proposed code bails when the correction field is negative, but then checks again whether it's negative. I think the second check is not necessary.

**Comment by andrew-elder** [comment]
Is correctionField value of 0 considered invalid as well. I would think so.

**Comment by PawelModrzejewski** [comment]
Zero is a valid value - the Grandmaster sends Followups with correctionField value of 0.

**Comment by christopher-s-hall** [comment]
I would rather this be a warning (actually exception print) and we continue to process rather than error and bail, since - at least in my reading - this isn't disallowed by the standard.

What about adding a port parameter that (dis)allows negative corrections and add that to the config file defaulting to disallow?

**Comment by PawelModrzejewski** [comment]
Sure, I'd be glad to work on it a little but from logical point of view, isn't a negative correction field an real erroneous situation indicator? I know that standard defines it as a singed integer value but when I'm reading 802.1AS-2011, for example section 11.1.3 (Transport of time-synchronization information), where it's defined as the sum of: a) the propagation delay and b) (...) the residence time, I have a feeling that in a good situations with linear time base it just can't be negative ... Besides that, sporadically daemon_cl's PID controller loop gets really unstable after processing such message and it was the main reason behind that PR.
Anyway I'll try to add suggested changes within a few days.

**Comment by andrew-elder** [comment]
Thanks for the comments everyone. @PawelModrzejewski - I'm assuming the "ball is in your court" for the moment!

**Comment by PawelModrzejewski** [comment]
-N' command line option added and 'allowNegativeCorrectionField' into gptp_cfg.ini

File: linux/src/daemon_cl.cpp [hunk 1]
```diff
@@ -429,6 +434,10 @@ int main(int argc, char **argv)
 			{
 				ether_phy_delay = iniParser.getPhyDelay();
 			}
+
+			portInit.allowNegativeCorrField = iniParser.getAllowNegativeCorrField();
+			GPTP_LOG_INFO("SyncFollowUp with negative correction field: %s",
+						  portInit.allowNegativeCorrField ? "permitted" : "forbiden");
```
Comment (andrew-elder) [inline linux/src/daemon_cl.cpp]: typo, should read "forbidden"

Comment (christopher-s-hall) [reply linux/src/daemon_cl.cpp]: It looks good to me. I would 'promote' the bail message to an ERROR and maybe direct the user to the config file to allow. Also, demote the WARNING to INFO. I want the 'user' to be able to suppress the message without suppressing  all warnings or modifying the code.

File: common/ptp_message.cpp [hunk 2]
```diff
@@ -1049,6 +1049,20 @@ void PTPMessageFollowUp::processMessage
...
+	if( correctionField < 0 )
+	{
+		if( port->getAllowNegativeCorrField() )
+		{
+			GPTP_LOG_WARNING
+					( "Received Follow Up with negative correctionField: %Ld", correctionField );
+		}
+		else
+		{
+			GPTP_LOG_EXCEPTION
```
Comment (christopher-s-hall) [inline common/ptp_message.cpp]:
```suggestion
			GPTP_LOG_ERROR
```

File: common/ptp_message.cpp [hunk 3]
```diff
@@ -1049,6 +1049,20 @@ void PTPMessageFollowUp::processMessage
 	master_local_freq_offset /= port->getPeerRateOffset();
 
 	correctionField /= 1 << 16;
+	if( correctionField < 0 )
+	{
+		if( port->getAllowNegativeCorrField() )
+		{
+			GPTP_LOG_WARNING
```
Comment (christopher-s-hall) [inline common/ptp_message.cpp]:
```suggestion
			GPTP_LOG_INFO
```

**Comment by andrew-elder** [comment]
@christopher-s-hall - I think this PR is now ok to be merged. Do you agree?

**Comment by christopher-s-hall** [comment]
Looks good

**Comment by andrew-elder** [comment]
Thanks Chris!

**Comment by PawelModrzejewski** [comment]
Thanks Guys for reviewing!

**Pull request merged**